### PR TITLE
Improve primitive_by_shape and an_sn_by_shape

### PR DESCRIPTION
**Pull request by fingolfin**
Add docstrings, explain why they are correct, add tests

### CODE REVIEW

File: experimental/GaloisGrp/GaloisGrp.jl [hunk 1]
```diff
@@ -1783,35 +1783,105 @@ function order_from_shape(ct::Set{CycleType}, n)
...
 
 
-#https://mathoverflow.net/questions/286316/recognition-of-symmetric-groups-in-gap
-#show primitivity by finding a permutation which has a p cycle for n/2<p<n−1
+# Given a set of cycle types of a group which acts transitively on n points,
+# this function returns `true` if the group must be primitive. Otherwise
+# `false` is returned (which does *not* imply that the group is imprimitive).
+#
+# See Theorem 10.2.1 on page 225 in Ákos Seress, Permutation group
+# algorithms, Cambridge Tracts in Mathematics, 152.
```
Comment (ThomasBreuer) [inline experimental/GaloisGrp/GaloisGrp.jl]: The book is worth to be added to the Oscar References.

File: experimental/GaloisGrp/GaloisGrp.jl [hunk 2]
```diff
@@ -1783,35 +1783,105 @@ function order_from_shape(ct::Set{CycleType}, n)
...
+
   return false
 end
 
+
+@doc Markdown.doc"""
+    an_sn_by_shape(ct::Set{CycleType}, n::Int)
+
+Return `true` if a transitive group $G \leq Sym(n)$ which contains permutations
+matching the cycle types in `ct` contains $Alt(n)$.
```
Comment (ThomasBreuer) [inline experimental/GaloisGrp/GaloisGrp.jl]: and `false` otherwise?

File: experimental/GaloisGrp/GaloisGrp.jl [hunk 3]
```diff
@@ -1783,35 +1783,105 @@ function order_from_shape(ct::Set{CycleType}, n)
...
+  # See e.g. Corollary 10.2.2 on page 225 in Ákos Seress, Permutation group
+  # algorithms, Cambridge Tracts in Mathematics, 152.
+  #
+  # Note that this condition is sufficient but not necessary. Indeed for n <= 7
+  # no such p exists, which is why we have to treat these degrees separately above.
+
+  # multiply all candidate primes
+if true
+  lp = prod(fmpz, PrimesSet(div(n, 2)+1, n-3))
+  return any(x->any(y ->gcd(y[1], lp) > 1, x.s), ct)
```
Comment (ThomasBreuer) [inline experimental/GaloisGrp/GaloisGrp.jl]: Why first form a product of primes and then check for a nontrivial gcd with this product? Is it not better to test whether one of the primes divides? (And why `if true`?)

**Review by ThomasBreuer — APPROVED** [review]
This is a nice improvement.
I have left a few minor comments.

File: experimental/GaloisGrp/GaloisGrp.jl (hunk 1 above)
Comment (fingolfin) [reply experimental/GaloisGrp/GaloisGrp.jl]: done

File: experimental/GaloisGrp/GaloisGrp.jl (hunk 2 above)
Comment (fingolfin) [reply experimental/GaloisGrp/GaloisGrp.jl]: done

File: experimental/GaloisGrp/GaloisGrp.jl (hunk 3 above)
Comment (fingolfin) [reply experimental/GaloisGrp/GaloisGrp.jl]: This was an optimization (?) already present in the code when I started to modify it. But I now did some benchmarks, and at least for small degrees using `is_prime` is actually faster (I first exclude values in the wrong range)

**Pull request merged**